source/language/types.rst

```diff
@@ -119,13 +119,24 @@ the quantum memory.
 Physical Qubits
 ~~~~~~~~~~~~~~~
 
-While program qubits can be named, hardware qubits are referenced only
+During quantum compilation it is critical to understand if an operation is being applied
+to an "virtual qubit" (a qubit pre-assignment) or a "physical" hardware qubit
+within a device. In most hardware implementations no two physical qubits are alike
+and reassigning physical qubits will change the impact of the quantum program on the
+system. For this reason OpenQASM specially identifies "physical qubits".
+
+
+Physical qubits are referenced only
 by the syntax ``$[NUM]``. For an ``n`` qubit system, we have physical qubit
 references given by ``$0``, ``$1``, ..., ``$n-1``. These qubit types are
 used in lower parts of the compilation stack when emitting physical
-circuits. Physical qubits must not be declared and they are, as all the qubits, global variables.
+circuits. For representation purposes physical qubits may be declared, eg.,
+``qubit $0;``. Implementations may also choose to represent available physical
+qubits implicitly.
 
 .. code-block::
+   // Declare physical qubit 0
+   qubit $0;
 
    // Declare a qubit
    qubit gamma;
```

Comment:
There is not much semantic difference between declaring `qubit $0` and just using `$0` in gates and so forth, at least according the the proposal and our thinking. So the possible collision between `gamma` and `$0` or `$1` seems to be there even as the spec stands now.

We might be informed by C:
```
register int x = 1;
int y = 2;
```
Presumably you can be sure that initializing `y` to `2` did not clobber the value in `x`. We might say the following:
> The set of physical qubits referred to in the program will be removed from the set of physical qubits that the virtual qubits will be mapped to.

Reply:
I explained more in my other comments, but briefly, I believe this:

> the set of physical qubits that the virtual qubits will be mapped to

to be a false premise.  In a general quantum program, there is no one-to-one mapping from virtual qubits to physical qubits - the mapping changes over the course of the execution.

(Semi-unrelated: `register int` doesn't "allocate" a register or anything like that, and it still has a type `int` - it's pretty much just a usage hint to a classical compiler for how hot the variable is.  Classical architectures also don't suffer the same problems; you can freely copy and move classical values from registers to RAM, but there's no equivalent to that here.  I was contending in other comments that `$0` and `q` are differently typed "data" entirely.)

Reply:
> In a general quantum program, there is no one-to-one mapping from virtual qubits to physical qubits - the mapping changes over the course of the execution.

The same exact thing happens in a "classical" program. There is no one-to-one mapping between variables and their values on the one side, and hardware registers on the other side. This mapping can - and does - change over the course of the execution, even though it may _appear_ that it doesn't. See _alias analysis_ as a prime example.
 
> I was contending in other comments that `$0` and `q` are differently typed "data" entirely.

That is already enforced by the compiler given the _special case identifier_ that can only be used for physical qubits. Physical qubit identifiers must begin with `$`. `$` is not allowed in any other identifier.